tests/asynch/waker.rs

```diff
@@ -49,3 +52,78 @@ fn wake_after_finish() {
         None,
     )
 }
+
+// Test that we can pass wakers across threads and have them work correctly
+#[test]
+fn wake_during_poll() {
+    check_dfs(
+        || {
+            let waker: Arc<Mutex<Option<Waker>>> = Arc::new(Mutex::new(None));
+            let waker_clone = Arc::clone(&waker);
+            let signal = Arc::new(AtomicBool::new(false));
+            let signal_clone = Arc::clone(&signal);
+
+            // This thread might invoke `wake` before the other task finishes running a single
+            // invocation of `poll`. If that happens, that task must not be blocked.
+            thread::spawn(move || {
+                signal_clone.store(true, Ordering::SeqCst);
+
+                if let Some(waker) = waker_clone.lock().unwrap().take() {
+                    waker.wake();
+                }
+            });
+
+            asynch::block_on(poll_fn(move |cx| {
+                *waker.lock().unwrap() = Some(cx.waker().clone());
+
+                if signal.load(Ordering::SeqCst) {
+                    Poll::Ready(())
+                } else {
+                    Poll::Pending
+                }
+            }));
+        },
+        None,
+    );
+}
+
+// Test that a waker invocation doesn't unblock a task that is blocked due to synchronization
+// operations
+#[test]
+fn wake_during_blocked_poll() {
+    check_dfs(
+        || {
+            let waker: Arc<Mutex<Option<Waker>>> = Arc::new(Mutex::new(None));
+            let waker_clone = Arc::clone(&waker);
+            let counter = Arc::new(Mutex::new(0));
+            let counter_clone = Arc::clone(&counter);
+
+            thread::spawn(move || {
+                let mut counter = counter_clone.lock().unwrap();
+                thread::yield_now();
+                *counter += 1;
+            });
+
+            // If this `wake()` invocation happens while the thread above holds the `counter` lock
+            // and the `block_on` task below is blocked waiting to acquire that same lock, then
+            // `wake` must not unblock the `block_on` task. That is, `wake` should prevent the task
+            // from being blocked *the next time it returns Pending*, not just any time it is
+            // blocked.
+            thread::spawn(move || {
+                if let Some(waker) = waker_clone.lock().unwrap().take() {
+                    waker.wake();
```

Comment:
Add a check (global var outside Shuttle) to ensure that we explore a run in which this line is executed?